File: AoC2024/dart/day11.dart
// --- Day 11: Plutonian Pebbles ---
import 'dart:io';

final test1 = [125, 17];

int part1(List<num> line, int blink) {
  var current = [...line];
  for (var i = 0; i < blink; i++) {
    print("Iteration $i");
    var new_line = <num>[];
    for (final value in current) {
      if (value == 0) {
        new_line.add(1);
      } else if (value.toString().length % 2 == 0) {
        final String vstr = value.toString();
        new_line.add(num.parse(vstr.substring(0, vstr.length ~/ 2)));
        new_line.add(num.parse(vstr.substring(vstr.length ~/ 2, vstr.length)));
      } else {
        new_line.add(value*2024);
      }
    }
    print("new: ${new_line.length}");
    current = [...new_line];
  }
  return current.length;
}



int part2(List<num> line, int blink) {
  print(line.map((e) => e.toString()).join('\n'));
  final c = File('current.txt').writeAsStringSync(line.map((e) => e.toString()).join('\n'));
  for (var i = 0; i < blink; i++) {
    print("*** blink $i ***");
    // final File('new.txt').createSync();
    final new_line = File('new.txt');
    final current = File('current.txt');
    for (final vstr in current.readAsLinesSync()) {
      int value = int.parse(vstr);
      String str = value.toString();
      // print(value);
      if (value == 0) {
        // print('$value -> 1');
        new_line.writeAsStringSync('1\n', mode: FileMode.append);
      } else if (str.length % 2 == 0) {
        // print('$value -> /2 /2');
        new_line.writeAsStringSync('${str.substring(0, str.length ~/ 2)}\n', mode: FileMode.append);
        new_line.writeAsStringSync('${str.substring(vstr.length ~/ 2, str.length)}\n', mode: FileMode.append);
      } else {
        // print('$value -> *2024');
        new_line.writeAsStringSync('${value*2024}\n', mode: FileMode.append);
      }
    }
    current.deleteSync();
    new_line.copySync('current.txt');
    new_line.deleteSync();
  }
  final current = File('current.txt');
  return current.readAsLinesSync().length;
}

void main() {
  // print ("test 1 by 6 iteration: ${part1(test1, 6)}");
  // print ("test 1 by 25 iteration: ${part1(test1, 25)}");

  String input = File('../datas/day11.txt').readAsStringSync();
  var list = List<num>.from(input.split(' ').map((e) => num.parse(e)));
  // print ("Part 1 input by 25 iteration: ${part1(list, 25)}");
  print ("Part 2 input by 75 iteration: ${part2(list, 75)}");
}

File: AoC2024/python/day11.py
# --- day11 ---
import utils
import os


def _rule(v: int) -> list[int]:
    vs = str(v)
    if v == 0:
        return [1]
    elif len(vs) % 2 == 0:
        return [int(f"{vs[:len(vs)//2]}"), int(f"{vs[len(vs)//2:]}")]
    else:
        return [v * 2024]


def part2(line: list, blick: int) -> int:
    current = {i: 1 for i in line}
    for b in range(blick):
        new = dict()
        for n, v in current.items():
            r = _rule(n)
            for i in r:
                if i in new:
                    new[i] += 1 * v
                else:
                    new[i] = 1 * v
        current = {k: v for k, v in new.items()}
        # print("blick", b + 1, current)
    return sum(current.values())


# print(part2([125, 17], 6))
# print(part2([125, 17], 25))
with open(os.path.join("..", "datas", "day11.txt")) as f:
    line = [int(i) for i in f.read().strip().split(" ")]
print("Day 11 Part2:", part2(line, 75))
